### PR TITLE
Added blindex tokens to rsk chain

### PR DESCRIPTION
**Pull request by MichalSzreder**

### CODE REVIEW

File: test/pages/OverviewPage.js [hunk 1]
```diff
@@ -199,6 +199,9 @@ class OverviewPage {
 
       case 'SOV':
       case 'FISH':
+      case 'BDX':
+      case 'BDUS':
+      case 'BDEU':
```
Comment (kazazor) [inline test/pages/OverviewPage.js:204]: How about XUSD & ETHs? Don't we need them as well? this applies to multiple places.

File: src/swaps/blindex/info.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "title": "Blindex AMM Swaps",
+  "description": "Popular DEX for Ethereum tokens only.",
+  "pros": ["High liquidity", "Many pairs", "Fast"],
+  "cons": ["Slippage", "Ethereum tokens only", "Few pairs"],
+  "fees": ["0.3% liquidity provider fee", "Slippage (up to 2% in Liquality)"]
+}
```
Comment (kazazor) [inline src/swaps/blindex/info.json]:
```suggestion
{
  "title": "Blindex AMM Swaps",
  "description": "Popular DEX for ERC20 tokens.",
  "pros": ["High liquidity", "Many pairs", "Fast"],
  "cons": ["ERC20 tokens only"],
  "fees": ["0.3% liquidity provider fee"]
}
```

File: src/swaps/blindex/BlindexConfig.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    { address: '0x15f2F01159a73A56a7149096942Ae4e2c019cbEf', symbol: 'UNI-V2', decimals: 18 },
+    { address: '0x3e64c5DC35546D6feF70346A395ed8E21C4A5247', symbol: 'UNI-V2', decimals: 18 },
+    { address: '0x2a5DC95E2F3150Ed52D69aD885EF6844dBED75C8', symbol: 'UNI-V2', decimals: 18 },
+    { address: '0x00ffA27B58312A06dbc4aEd883d5F2EA924b2F89', symbol: 'UNI-V2', decimals: 18 },
+    { address: '0xAB53CC054Ac0722aB0cBF346495E64c8C8e0c346', symbol: 'UNI-V2', decimals: 18 },
+    { address: '0xdc7F0077E1921dBBCf8D042EE49184716301EDFE', symbol: 'UNI-V2', decimals: 18 },
+    { address: '0x39Eb6A2601aC29526E37EA60a531223AB0679873', symbol: 'UNI-V2', decimals: 18 },
+    { address: '0x6a11d6af629bBbf310B82312E5fbd2dcc997a781', symbol: 'UNI-V2', decimals: 18 },
+    { address: '0x4B481D8024B55263455AD3329e8be62C03A3F315', symbol: 'UNI-V2', decimals: 18 },
+    { address: '0x2De59550090353D89fE36c50482234D9B7DB2e3B', symbol: 'UNI-V2', decimals: 18 }
```
Comment (kazazor) [inline src/swaps/blindex/BlindexConfig.js]: Why do we need all the LP tokens as well?

**Review by kazazor — COMMENTED** [review]
Will continue my review tomorrow, but wanted to get the ball rolling for now.

File: src/swaps/blindex/BlindexConfig.js (hunk 3 above)
Comment (MichalSzreder) [reply src/swaps/blindex/BlindexConfig.js:91]: Will change it to just wrbtc address and list of possible pair symbols.

File: src/swaps/blindex/BestRouteService.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
+import { blindexConfig } from './BlindexConfig'
+
+export class BestRouteService {
+  availableLinks = []
+
+  constructor() {
+    const availableLinks = []
```
Comment (kazazor) [inline src/swaps/blindex/BestRouteService.js]: Redundant variable. You can just use `this.availableLinks` instead.

File: src/swaps/blindex/BestRouteService.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+    for (const pair of blindexConfig.SWAPS) {
+      availableLinks.push({ from: pair.token0.toLowerCase(), to: pair.token1.toLowerCase() })
+      availableLinks.push({ from: pair.token1.toLowerCase(), to: pair.token0.toLowerCase() })
+    }
```
Comment (kazazor) [inline src/swaps/blindex/BestRouteService.js]: Why store the tokens in lower case? Wouldn't it work also without it?

File: src/swaps/blindex/BestRouteService.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+      for (const link2 of this.availableLinks) {
+        if (link1.to !== link2.from) {
+          continue
+        }
+        if (link2.to !== addressOut) {
+          continue
+        }
+
+        midTokens.push(link1.to)
+      }
```
Comment (kazazor) [inline src/swaps/blindex/BestRouteService.js]:
```suggestion
      if (link1.from === addressIn) {
          for (const link2 of this.availableLinks) {
            if (link1.to === link2.from && link2.to === addressOut) {
                midTokens.push(link1.to)
            }
          }
       }
```

Same result with less code

File: src/swaps/blindex/BestRouteService.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  async getRoutesWithAmount(router, routes, amount, invokeGetAmountsIn) {
+    const routesPrices = []
+    for (const route of routes) {
+      let amounts
+      try {
+        amounts = invokeGetAmountsIn
+          ? await router.getAmountsIn(amount, route)
+          : await router.getAmountsOut(amount, route)
+      } catch (e) {
+        continue // handle unsupported paths like [wrbtc -> eths -> bdx]
```
Comment (kazazor) [inline src/swaps/blindex/BestRouteService.js]: How can that happen? From what I see in `generateRoutes` you must have a path from BTC to ETH for that to happen, isn't it?

File: src/swaps/blindex/BestRouteService.js [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    }
+    const routes = [...midTokens.map((x) => [addressIn, x, addressOut])]
+    if (this.availableLinks.some((link) => link.from === addressIn && link.to === addressOut)) {
+      routes.push([addressIn, addressOut])
+    }
+
+    return routes
+  }
+
+  async chooseBestRoute(routesInfos, invokeGetAmountsIn) {
```
Comment (kazazor) [inline src/swaps/blindex/BestRouteService.js]: NIT
so the naming will remain the same for better redability
```suggestion
  async chooseBestRoute(allRoutesWithAmounts, invokeGetAmountsIn) {
```

File: src/swaps/blindex/BestRouteService.js [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    const bestPath = routesInfos.reduce((prev, current) => {
+      const selectedRouteInfo = invokeGetAmountsIn
+        ? prev.finalAmount.lt(current.finalAmount) ||
+          (prev.finalAmount.eq(current.finalAmount) && prev.route.length < current.route.length)
```
Comment (kazazor) [inline src/swaps/blindex/BestRouteService.js]:
```suggestion
        ? prev.finalAmount.lte(current.finalAmount) && prev.route.length < current.route.length)
```

File: src/swaps/blindex/BestRouteService.js [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+          ? prev
+          : current
+        : prev.finalAmount.gt(current.finalAmount) ||
+          (prev.finalAmount.eq(current.finalAmount) && prev.route.length < current.route.length)
```
Comment (kazazor) [inline src/swaps/blindex/BestRouteService.js]:
```suggestion
        : prev.finalAmount.gte(current.finalAmount) && prev.route.length < current.route.length)
```

File: src/swaps/blindex/BestRouteService.js [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    if (this.availableLinks.some((link) => link.from === addressIn && link.to === addressOut)) {
+      routes.push([addressIn, addressOut])
+    }
+
+    return routes
+  }
+
+  async chooseBestRoute(routesInfos, invokeGetAmountsIn) {
+    const bestPath = routesInfos.reduce((prev, current) => {
+      const selectedRouteInfo = invokeGetAmountsIn
```
Comment (kazazor) [inline src/swaps/blindex/BestRouteService.js]: Please let's use `if`/`else` instead of `?` `:` when it comes to `invokeGetAmountsIn`.. Too much of these and it's hard to read

File: src/swaps/blindex/BestRouteService.js (hunk 9 above)
Comment (kazazor) [inline src/swaps/blindex/BestRouteService.js]: Not sure I understand this part:
`prev.route.length < current.route.length`

Can you explain why does it matter when choosing the best route?
From what I see here, basically, if the amount we get in a `direct` route is better than an indirect route, we're still not choosing it because the route length is shorter.. Which seems wrong..

What am I missing here?

File: src/swaps/blindex/BlindexSwapProvider.js [hunk 12]
```diff
@@ -0,0 +1,509 @@
+import BN from 'bignumber.js'
```
Comment (kazazor) [inline src/swaps/blindex/BlindexSwapProvider.js:1]: I see you're using both `bignumber.js` and `ethers.BigNumber`
do we really need both?

**Review by kazazor — CHANGES_REQUESTED** [review]
I did not finish my review yet (jumping into meetings here), but I wanted to already share my comments.
I only need to still review the BlindexSwapProvider.js file

File: src/swaps/blindex/BestRouteService.js (hunk 5 above)
Comment (MichalSzreder) [reply src/swaps/blindex/BestRouteService.js]: Tokens are compared later in _generateRoutes_ method and I think lowercasing them here on initialization makes _generateRoutes_ easier to read.

File: src/swaps/blindex/BestRouteService.js (hunk 6 above)
Comment (MichalSzreder) [reply src/swaps/blindex/BestRouteService.js]: Yeah, we have the same code in frontend app.
We might create some npm package to keep one codebase.
WDYT?

File: src/swaps/blindex/BestRouteService.js (hunk 7 above)
Comment (MichalSzreder) [reply src/swaps/blindex/BestRouteService.js]: Yeah, it would step out of the inner loop on 
`        
        if (link1.to !== link2.from) {
          continue
        }
`
It's @AndrzejGisWideLab comment originating to smart routing poc.
I could remove whole try-catch statement.
@AndrzejGisWideLab is there something we miss?

File: src/swaps/blindex/BlindexSwapProvider.js (hunk 12 above)
Comment (MichalSzreder) [reply src/swaps/blindex/BlindexSwapProvider.js:1]: Liquality uses mostly bignumber.js internally (e.g. data returned from overriden SwapProvider methods return bignumber.js objects). BestRouteService and UniswapSwapProvider, for which I copied some code uses ethers.BigNumber. BestRouteService just for calculations, and we might change it, but code from UniswapSwapProvider uses ether.BigNumber to use _toHexString_ method for _encodeFunctionData_ invocation.

TLDR
I would keep using both.

File: src/swaps/blindex/BestRouteService.js (hunk 9 above)
Comment (MichalSzreder) [reply src/swaps/blindex/BestRouteService.js]: prev.finalAmount.lt(current.finalAmount) ||
(prev.finalAmount.eq(current.finalAmount) && prev.route.length < current.route.length)
          ? prev
          : current
          
means

if prev route has better(smaller) amount
or
prev route has the same amount and is shorter
then
take prev route
otherwise take current

I thought it would be better for users to choose shorter route for lower fee (0.3% instead 0.6%) in case the amounts in one or more routes are equal. Even if the chance of that happening is very small.

File: src/swaps/blindex/BestRouteService.js (hunk 6 above)
Comment (kazazor) [reply src/swaps/blindex/BestRouteService.js]: In general that could have been good, but for now let's not do it and just change it here

File: src/swaps/blindex/BestRouteService.js (hunk 5 above)
Comment (kazazor) [reply src/swaps/blindex/BestRouteService.js]: Why would it make it easier to read? You can just not use `toLowerCase` at all here and there, right?

File: src/swaps/blindex/BestRouteService.js (hunk 9 above)
Comment (kazazor) [reply src/swaps/blindex/BestRouteService.js]: OK got it right now, I think everything got mixed to me when. looked at it before due to all the use of `?` and `:`, would be great making it a bit more readable

Comment (MichalSzreder) [reply src/swaps/blindex/BestRouteService.js]: I hope if/else change would do the job

File: src/swaps/blindex/BestRouteService.js [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+    const routes = [...midTokens.map((x) => [addressIn, x, addressOut])]
+    if (this.availableLinks.some((link) => link.from === addressIn && link.to === addressOut)) {
+      routes.push([addressIn, addressOut])
+    }
+
+    return routes
+  }
+
+  async chooseBestRoute(allRoutesWithAmounts, invokeGetAmountsIn) {
+    const bestPath = routesInfos.reduce((prev, current) => {
```
Comment (tkeidar) [inline src/swaps/blindex/BestRouteService.js]: `routesInfos` is not defined here, did you mean `allRoutesWithAmounts`?

not sure how could it ever work this way...

File: test/e2e/import_wallet.spec.js [hunk 14]
```diff
@@ -171,7 +171,7 @@ describe('Import wallet-["MAINNET"]', async () => {
     ).equals(ethAddress)
 
     // Check RSK & ERC20 tokens
-    const rskTokens = ['RBTC', 'SOV', 'FISH']
+    const rskTokens = ['RBTC', 'SOV', 'FISH', 'BDX', 'BDUS', 'BDEU']
```
Comment (ghost) [inline test/e2e/import_wallet.spec.js]: XUSD & ETHs are missing also here

File: src/views/Swap/SwapInfo.vue [hunk 15]
```diff
@@ -32,7 +32,9 @@ export default {
     showSlippageMessage() {
       const providerType = getSwapProviderConfig(this.activeNetwork, this.quote.provider).type
       return (
-        providerType !== SwapProviderType.LIQUALITY && providerType !== SwapProviderType.FASTBTC
+        providerType !== SwapProviderType.LIQUALITY &&
+        providerType !== SwapProviderType.FASTBTC &&
+        providerType !== SwapProviderType.BLINDEX
```
Comment (ghost) [inline src/views/Swap/SwapInfo.vue:37]: why no slippage for blindex? It looks like most providers show slippage message

File: src/swaps/blindex/BestRouteService.js (hunk 7 above)
Comment (ghost) [reply src/swaps/blindex/BestRouteService.js]: @kazazor @MichalSzreder It depends on how you generate the routes. I think originally it was just all combinations of tokens. Later it was changed to be the combinations of matching pairs so probably the catch is unnecessary

File: src/swaps/blindex/BestRouteService.js (hunk 5 above)
Comment (ghost) [reply src/swaps/blindex/BestRouteService.js]: @kazazor probably @MichalSzreder is referring to the fact that no `toLowerCase` calls are needed in `generateRoutes`. 

@MichalSzreder  I think @kazazor is right. No reason to store transformed data. It's better to transform the addresses when you explicitly need them to match other lower cased addresses.

File: src/swaps/blindex/BlindexSwapProvider.js [hunk 16]
```diff
@@ -0,0 +1,509 @@
...
+      assetData.contractAddress,
+      assetData.decimals,
+      assetData.code,
+      assetData.name
+    )
+  }
+
+  getMinimumOutput(outputAmount) {
+    const slippageTolerance = new Percent('200', '10000') // 2%
+    const slippageAdjustedAmountOut = new Fraction(JSBI.BigInt(1))
```
Comment (ghost) [inline src/swaps/blindex/BlindexSwapProvider.js:70]: Does it come from BingNumber family or is it another library?

File: src/swaps/blindex/BestRouteService.js (hunk 13 above)
Comment (kazazor) [reply src/swaps/blindex/BestRouteService.js]: I believe he just approved my suggestion for changing the name but did not change the other places that uses the variable.. right?

File: src/swaps/blindex/BestRouteService.js (hunk 7 above)
Comment (kazazor) [reply src/swaps/blindex/BestRouteService.js]: Tom, why did you resolved this comment?

Comment (tkeidar) [reply src/swaps/blindex/BestRouteService.js]: Probably by mistake

File: src/swaps/blindex/BestRouteService.js (hunk 13 above)
Comment (tkeidar) [reply src/swaps/blindex/BestRouteService.js]: Could be. 
@MichalSzreder is there any chance that's the reason I'm not able to perform a swap?

File: src/swaps/blindex/BestRouteService.js (hunk 7 above)
Comment (kazazor) [reply src/swaps/blindex/BestRouteService.js]: Thanks @AndrzejGisWideLab , so yeah let's remove it

File: src/swaps/blindex/BlindexSwapProvider.js [hunk 17]
```diff
@@ -0,0 +1,509 @@
...
+const RSK_RPC_URL = 'https://public-node.rsk.co'
+
+class BlindexSwapProvider extends SwapProvider {
+  constructor(config) {
+    super(config)
+    this._apiCache = {}
+    this.bestRouteService = new BestRouteService()
+    this.wrbtcAddress = blindexConfig.ERC20_INFO.find((x) => x.symbol === 'WRBTC').address
+    this.nativeTokenSymbol = chains[ChainId.Rootstock].nativeAsset //RBTC
+    this.availableTokenSymbols = blindexConfig.ERC20_INFO.filter((entry) => entry !== 'UNI-V2').map(
```
Comment (kazazor) [inline src/swaps/blindex/BlindexSwapProvider.js]: Once you'll remove the UNI tokens, this could be removed
```suggestion
    this.availableTokenSymbols = blindexConfig.ERC20_INFO.map(
```

File: src/swaps/blindex/BlindexSwapProvider.js [hunk 18]
```diff
@@ -0,0 +1,509 @@
...
+      (entry) => (entry.symbol === 'WRBTC' ? this.nativeTokenSymbol : entry.symbol)
+    )
+  }
+
+  async getSupportedPairs() {
+    return []
+  }
```
Comment (kazazor) [inline src/swaps/blindex/BlindexSwapProvider.js:39]: Is this here by mistake? I see it's empty and not being used

File: src/swaps/blindex/BlindexSwapProvider.js [hunk 19]
```diff
@@ -0,0 +1,509 @@
...
+  async newSwap({ network, walletId, quote }) {
+    const approvalRequired = isERC20(quote.from)
+    const updates = approvalRequired
+      ? await this.approveTokens({ network, walletId, quote })
+      : await this.sendSwap({ network, walletId, quote })
+
+    return {
+      id: uuidv4(),
+      fee: quote.fee,
+      slippage: 200,
```
Comment (kazazor) [inline src/swaps/blindex/BlindexSwapProvider.js]: This is the second time we're defining the slippage. Can we have it in one place and reuse it?
So if we'll change it in the future we won't forget to update all the places

File: src/swaps/blindex/BlindexSwapProvider.js [hunk 20]
```diff
@@ -0,0 +1,509 @@
...
+
+    return {
+      status: 'WAITING_FOR_SWAP_CONFIRMATIONS',
+      swapTx,
+      swapTxHash: swapTx.hash
+    }
+  }
+
+  getChecksumAddressFromAddress(address) {
+    return ethers.utils.getAddress(address.toLowerCase())
```
Comment (kazazor) [inline src/swaps/blindex/BlindexSwapProvider.js:145]: Do we need to use `toLowerCase` here?
I believe `ethers.utils.getAddress` makes sure we get the right address even without it

File: src/swaps/blindex/BestRouteService.js (hunk 13 above)
Comment (kazazor) [reply src/swaps/blindex/BestRouteService.js]: @tkeidar tried it offline. it was the reason for that, so just need to fix it.

**Review by kazazor — CHANGES_REQUESTED** [review]
I finished all my comments, overall really nice work!

File: src/swaps/blindex/BestRouteService.js (hunk 13 above)
Comment (tkeidar) [reply src/swaps/blindex/BestRouteService.js]: I just pushed a fix for this

File: src/assets/icons/assets/bdeu.svg [hunk 21]
```diff
@@ -0,0 +1,12 @@
+<svg width="36" height="36" viewBox="0 0 36 36" fill="none" xmlns="http://www.w3.org/2000/svg">
```
Comment (tkeidar) [inline src/assets/icons/assets/bdeu.svg:1]: XUSD and ETHs icons are also missing
image

**Review by tkeidar — CHANGES_REQUESTED** [review]
@MichalSzreder let's fix the current open comments and review this PR again

File: test/pages/OverviewPage.js (hunk 1 above)
Comment (MichalSzreder) [reply test/pages/OverviewPage.js:204]: done

File: src/swaps/blindex/BlindexConfig.js (hunk 3 above)
Comment (MichalSzreder) [reply src/swaps/blindex/BlindexConfig.js]: Removed LP tokens and decimals information

File: src/swaps/blindex/BestRouteService.js (hunk 4 above)
Comment (MichalSzreder) [reply src/swaps/blindex/BestRouteService.js]: done

File: src/swaps/blindex/BestRouteService.js (hunk 5 above)
Comment (MichalSzreder) [reply src/swaps/blindex/BestRouteService.js]: done

File: src/swaps/blindex/BestRouteService.js (hunk 6 above)
Comment (MichalSzreder) [reply src/swaps/blindex/BestRouteService.js]: done

File: src/swaps/blindex/BestRouteService.js (hunk 7 above)
Comment (MichalSzreder) [reply src/swaps/blindex/BestRouteService.js]: done

File: src/swaps/blindex/BestRouteService.js (hunk 11 above)
Comment (MichalSzreder) [reply src/swaps/blindex/BestRouteService.js]: done

File: test/e2e/import_wallet.spec.js (hunk 14 above)
Comment (MichalSzreder) [reply test/e2e/import_wallet.spec.js]: done

File: src/views/Swap/SwapInfo.vue (hunk 15 above)
Comment (MichalSzreder) [reply src/views/Swap/SwapInfo.vue:37]: This function only determines whether to show slippage message, not what that message is.
And the message is 'Max slippage is 0.5%'.
We might make it configurable and show 2% but I thought we may skip it for now (regarding the fact Liquality itself which has most of the swaps doesn't show it).
image

File: src/swaps/blindex/BlindexSwapProvider.js (hunk 16 above)
Comment (MichalSzreder) [reply src/swaps/blindex/BlindexSwapProvider.js:70]: It's a dependency of @uniswap/sdk-core, it's not bignumber, rather part of ECMAscript (was included in the standard in ES2020 version)

File: src/swaps/blindex/BlindexSwapProvider.js (hunk 17 above)
Comment (MichalSzreder) [reply src/swaps/blindex/BlindexSwapProvider.js]: done

File: src/swaps/blindex/BlindexSwapProvider.js (hunk 20 above)
Comment (MichalSzreder) [reply src/swaps/blindex/BlindexSwapProvider.js:145]: We need.
If given address is mixed-cased then getAddress has another check to verify if the address is equal to checksum address.
If its not it logs an error.
The config which I have copied from frontent environment.prod does not have checksummed addresses apparently.
image

File: src/assets/icons/assets/bdeu.svg (hunk 21 above)
Comment (MichalSzreder) [reply src/assets/icons/assets/bdeu.svg:1]: That's weird, they show up for me. Did you try rebuilding wallet?
The svg files though have content a bit different from other svgs, which already have been there. I may investigate that if this keeps occuring in the 2nd round of testing.

File: src/swaps/blindex/BlindexSwapProvider.js (hunk 19 above)
Comment (MichalSzreder) [reply src/swaps/blindex/BlindexSwapProvider.js]: done

File: src/swaps/blindex/BlindexSwapProvider.js (hunk 18 above)
Comment (MichalSzreder) [reply src/swaps/blindex/BlindexSwapProvider.js:39]: That's how almost every swap provider is implemented. The Liquality and FastBtc swap provider have them defined but as far as I can see the data is then saved in the application store as 'marketData' for later use.
And it is later used to feed data to Liquality swap provider. So I suppose it serves a purpose of caching the data for easier access.
image

File: src/assets/icons/assets/bdeu.svg (hunk 21 above)
Comment (tkeidar) [reply src/assets/icons/assets/bdeu.svg:1]: it's working now 👍

File: src/swaps/blindex/BlindexSwapProvider.js [hunk 22]
```diff
@@ -239,7 +238,7 @@ class BlindexSwapProvider extends SwapProvider {
     const blockHeight = await client.chain.getBlockHeight()
     const currentBlock = await client.chain.getBlockByNumber(blockHeight)
 
-    const path = [fromToken.address, toToken.address]
+    const path = this.bestRoute
```
Comment (tkeidar) [inline src/swaps/blindex/BlindexSwapProvider.js]: you can just use `this.bestRoute` instead of creating another variable

**Review by tkeidar — APPROVED** [review]

File: src/swaps/blindex/BlindexSwapProvider.js (hunk 22 above)
Comment (MichalSzreder) [reply src/swaps/blindex/BlindexSwapProvider.js]: done

**Review by kazazor — APPROVED** [review]
